File: photonlib-java-examples/aimandrange/src/main/java/frc/robot/subsystems/drivetrain/SwerveModule.java
/*
 * MIT License
 *
 * Copyright (c) PhotonVision
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package frc.robot.subsystems.drivetrain;

import static frc.robot.Constants.Swerve.*;

import edu.wpi.first.math.MathUtil;
import edu.wpi.first.math.controller.PIDController;
import edu.wpi.first.math.geometry.Rotation2d;
import edu.wpi.first.math.kinematics.SwerveModulePosition;
import edu.wpi.first.math.kinematics.SwerveModuleState;
import edu.wpi.first.math.util.Units;
import edu.wpi.first.wpilibj.Encoder;
import edu.wpi.first.wpilibj.RobotController;
import edu.wpi.first.wpilibj.motorcontrol.PWMSparkMax;
import edu.wpi.first.wpilibj.simulation.EncoderSim;
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;

public class SwerveModule {
    // --- Module Constants
    private final ModuleConstants moduleConstants;

    // --- Hardware
    private final PWMSparkMax driveMotor;
    private final Encoder driveEncoder;
    private final PWMSparkMax steerMotor;
    private final Encoder steerEncoder;

    // --- Control
    private SwerveModuleState desiredState = new SwerveModuleState();
    private boolean openLoop = false;

    // Simple PID feedback controllers run on the roborio
    private PIDController drivePidController = new PIDController(kDriveKP, kDriveKI, kDriveKD);
    // (A profiled steering PID controller may give better results by utilizing feedforward.)
    private PIDController steerPidController = new PIDController(kSteerKP, kSteerKI, kSteerKD);

    // --- Simulation
    private final EncoderSim driveEncoderSim;
    private double driveCurrentSim = 0;
    private final EncoderSim steerEncoderSim;
    private double steerCurrentSim = 0;

    public SwerveModule(ModuleConstants moduleConstants) {
        this.moduleConstants = moduleConstants;

        driveMotor = new PWMSparkMax(moduleConstants.driveMotorID);
        driveEncoder = new Encoder(moduleConstants.driveEncoderA, moduleConstants.driveEncoderB);
        driveEncoder.setDistancePerPulse(kDriveDistPerPulse);
        steerMotor = new PWMSparkMax(moduleConstants.steerMotorID);
        steerEncoder = new Encoder(moduleConstants.steerEncoderA, moduleConstants.steerEncoderB);
        steerEncoder.setDistancePerPulse(kSteerRadPerPulse);

        steerPidController.enableContinuousInput(-Math.PI, Math.PI);

        // --- Simulation
        driveEncoderSim = new EncoderSim(driveEncoder);
        steerEncoderSim = new EncoderSim(steerEncoder);
    }

    public void periodic() {
        // Perform PID feedback control to steer the module to the target angle
        double steerPid =
                steerPidController.calculate(
                        getAbsoluteHeading().getRadians(), desiredState.angle.getRadians());
        steerMotor.setVoltage(steerPid);

        // Use feedforward model to translate target drive velocities into voltages
        double driveFF = kDriveFF.calculate(desiredState.speedMetersPerSecond);
        double drivePid = 0;
        if (!openLoop) {
            // Perform PID feedback control to compensate for disturbances
            drivePid =
                    drivePidController.calculate(driveEncoder.getRate(), desiredState.speedMetersPerSecond);
        }

        driveMotor.setVoltage(driveFF + drivePid);
    }

    /**
     * Command this swerve module to the desired angle and velocity.
     *
     * @param steerInPlace If modules should steer to target angle when target velocity is 0
     */
    public void setDesiredState(
            SwerveModuleState desiredState, boolean openLoop, boolean steerInPlace) {
        Rotation2d currentRotation = getAbsoluteHeading();
        // Avoid turning more than 90 degrees by inverting speed on large angle changes
        desiredState.optimize(currentRotation);

        this.desiredState = desiredState;
    }

    /** Module heading reported by steering encoder. */
    public Rotation2d getAbsoluteHeading() {
        return Rotation2d.fromRadians(steerEncoder.getDistance());
    }

    /**
     * {@link SwerveModuleState} describing absolute module rotation and velocity in meters per
     * second.
     */
    public SwerveModuleState getState() {
        return new SwerveModuleState(driveEncoder.getRate(), getAbsoluteHeading());
    }

    /** {@link SwerveModulePosition} describing absolute module rotation and position in meters. */
    public SwerveModulePosition getPosition() {
        return new SwerveModulePosition(driveEncoder.getDistance(), getAbsoluteHeading());
    }

    /** Voltage of the drive motor */
    public double getDriveVoltage() {
        return driveMotor.get() * RobotController.getBatteryVoltage();
    }

    /** Voltage of the steer motor */
    public double getSteerVoltage() {
        return steerMotor.get() * RobotController.getBatteryVoltage();
    }

    /**
     * Constants about this module, like motor IDs, encoder angle offset, and translation from center.
     */
    public ModuleConstants getModuleConstants() {
        return moduleConstants;
    }

    public void log() {
        var state = getState();

        String table = "Module " + moduleConstants.moduleNum + "/";
        SmartDashboard.putNumber(
                table + "Steer Degrees", Math.toDegrees(MathUtil.angleModulus(state.angle.getRadians())));
        SmartDashboard.putNumber(
                table + "Steer Target Degrees", Math.toDegrees(steerPidController.getSetpoint()));
        SmartDashboard.putNumber(
                table + "Drive Velocity Feet", Units.metersToFeet(state.speedMetersPerSecond));
        SmartDashboard.putNumber(
                table + "Drive Velocity Target Feet",
                Units.metersToFeet(desiredState.speedMetersPerSecond));
        SmartDashboard.putNumber(table + "Drive Current", driveCurrentSim);
        SmartDashboard.putNumber(table + "Steer Current", steerCurrentSim);
    }

    // ----- Simulation

    public void simulationUpdate(
            double driveEncoderDist,
            double driveEncoderRate,
            double driveCurrent,
            double steerEncoderDist,
            double steerEncoderRate,
            double steerCurrent) {
        driveEncoderSim.setDistance(driveEncoderDist);
        driveEncoderSim.setRate(driveEncoderRate);
        this.driveCurrentSim = driveCurrent;
        steerEncoderSim.setDistance(steerEncoderDist);
        steerEncoderSim.setRate(steerEncoderRate);
        this.steerCurrentSim = steerCurrent;
    }

    public double getDriveCurrentSim() {
        return driveCurrentSim;
    }

    public double getSteerCurrentSim() {
        return steerCurrentSim;
    }
}

File: photonlib-cpp-examples/poseest/src/main/cpp/subsystems/SwerveModule.cpp
/*
 * MIT License
 *
 * Copyright (c) PhotonVision
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

#include "subsystems/SwerveModule.h"

#include <string>

#include <wpi/math/util/MathUtil.hpp>
#include <wpi/smartdashboard/SmartDashboard.hpp>
#include <wpi/system/RobotController.hpp>

SwerveModule::SwerveModule(const constants::Swerve::ModuleConstants& consts)
    : moduleConstants(consts),
      driveMotor(wpi::PWMSparkMax{moduleConstants.driveMotorId}),
      driveEncoder(wpi::Encoder{moduleConstants.driveEncoderA,
                                moduleConstants.driveEncoderB}),
      steerMotor(wpi::PWMSparkMax{moduleConstants.steerMotorId}),
      steerEncoder(wpi::Encoder{moduleConstants.steerEncoderA,
                                moduleConstants.steerEncoderB}),
      driveEncoderSim(driveEncoder),
      steerEncoderSim(steerEncoder) {
  driveEncoder.SetDistancePerPulse(
      constants::Swerve::kDriveDistPerPulse.to<double>());
  steerEncoder.SetDistancePerPulse(
      constants::Swerve::kSteerRadPerPulse.to<double>());
  steerPIDController.EnableContinuousInput(-std::numbers::pi, std::numbers::pi);
}

void SwerveModule::Periodic() {
  wpi::units::volt_t steerPID = wpi::units::volt_t{
      steerPIDController.Calculate(GetAbsoluteHeading().Radians().to<double>(),
                                   desiredState.angle.Radians().to<double>())};
  steerMotor.SetVoltage(steerPID);

  wpi::units::volt_t driveFF =
      constants::Swerve::kDriveFF.Calculate(desiredState.speed);
  wpi::units::volt_t drivePID{0};
  if (!openLoop) {
    drivePID = wpi::units::volt_t{drivePIDController.Calculate(
        driveEncoder.GetRate(), desiredState.speed.to<double>())};
  }
  driveMotor.SetVoltage(driveFF + drivePID);
}

void SwerveModule::SetDesiredState(wpi::math::SwerveModuleState newState,
                                   bool shouldBeOpenLoop, bool steerInPlace) {
  wpi::math::Rotation2d currentRotation = GetAbsoluteHeading();
  newState.Optimize(currentRotation);
  desiredState = newState;
}

wpi::math::Rotation2d SwerveModule::GetAbsoluteHeading() const {
  return wpi::math::Rotation2d{
      wpi::units::radian_t{steerEncoder.GetDistance()}};
}

wpi::math::SwerveModuleState SwerveModule::GetState() const {
  return wpi::math::SwerveModuleState{driveEncoder.GetRate() * 1_mps,
                                      GetAbsoluteHeading()};
}

wpi::math::SwerveModulePosition SwerveModule::GetPosition() const {
  return wpi::math::SwerveModulePosition{driveEncoder.GetDistance() * 1_m,
                                         GetAbsoluteHeading()};
}

wpi::units::volt_t SwerveModule::GetDriveVoltage() const {
  return driveMotor.Get() * wpi::RobotController::GetBatteryVoltage();
}

wpi::units::volt_t SwerveModule::GetSteerVoltage() const {
  return steerMotor.Get() * wpi::RobotController::GetBatteryVoltage();
}

wpi::units::ampere_t SwerveModule::GetDriveCurrentSim() const {
  return driveCurrentSim;
}

wpi::units::ampere_t SwerveModule::GetSteerCurrentSim() const {
  return steerCurrentSim;
}

constants::Swerve::ModuleConstants SwerveModule::GetModuleConstants() const {
  return moduleConstants;
}

void SwerveModule::Log() {
  wpi::math::SwerveModuleState state = GetState();

  std::string table =
      "Module " + std::to_string(moduleConstants.moduleNum) + "/";
  wpi::SmartDashboard::PutNumber(table + "Steer Degrees",
                                 wpi::math::AngleModulus(state.angle.Radians())
                                     .convert<wpi::units::degrees>()
                                     .to<double>());
  wpi::SmartDashboard::PutNumber(
      table + "Steer Target Degrees",
      wpi::units::radian_t{steerPIDController.GetSetpoint()}
          .convert<wpi::units::degrees>()
          .to<double>());
  wpi::SmartDashboard::PutNumber(
      table + "Drive Velocity Feet",
      state.speed.convert<wpi::units::feet_per_second>().to<double>());
  wpi::SmartDashboard::PutNumber(
      table + "Drive Velocity Target Feet",
      desiredState.speed.convert<wpi::units::feet_per_second>().to<double>());
  wpi::SmartDashboard::PutNumber(table + "Drive Current",
                                 driveCurrentSim.to<double>());
  wpi::SmartDashboard::PutNumber(table + "Steer Current",
                                 steerCurrentSim.to<double>());
}

void SwerveModule::SimulationUpdate(
    wpi::units::meter_t driveEncoderDist,
    wpi::units::meters_per_second_t driveEncoderRate,
    wpi::units::ampere_t driveCurrent, wpi::units::radian_t steerEncoderDist,
    wpi::units::radians_per_second_t steerEncoderRate,
    wpi::units::ampere_t steerCurrent) {
  driveEncoderSim.SetDistance(driveEncoderDist.to<double>());
  driveEncoderSim.SetRate(driveEncoderRate.to<double>());
  driveCurrentSim = driveCurrent;
  steerEncoderSim.SetDistance(steerEncoderDist.to<double>());
  steerEncoderSim.SetRate(steerEncoderRate.to<double>());
  steerCurrentSim = steerCurrent;
}
